File: bada_ruby/lib/bada/transformer/vision.rb
# frozen_string_literal: true

require_relative "tensor"
require_relative "model"

module Bada
  module Transformer
    # Vision head — an image-processing Transformer (ViT-style) on top of the
    # same manifold-gauge Encoder. A grid of intensities is cut into patches,
    # each patch is linearly projected to a token embedding, the encoder mixes
    # them under the gamma-manifold gauge, and each patch is projected back to an
    # intensity — the reconstructed "mental image" (対象の脳内に浮かぶ画像).
    class Vision
      T = Tensor

      def initialize(encoder:, grid: 8, patch: 2, seed: 0x71501)
        @encoder = encoder
        @grid = grid
        @patch = patch
        @per_side = grid / patch
        @n_patches = @per_side * @per_side
        @patch_dim = patch * patch
        rng = T::PRNG.new(seed)
        @proj_in = T.randmat(@patch_dim, encoder.d_model, rng)   # patch -> token
        @proj_out = T.randmat(encoder.d_model, 1, rng)           # token -> intensity
      end

      attr_reader :grid, :n_patches

      # image: grid x grid array of floats in [0,1]. Returns a reconstructed
      # grid x grid array in [0,1] (the transformer's rendered mental image).
      def process(image)
        patches = patchify(image)                 # n_patches x patch_dim
        tokens = T.matmul(patches, @proj_in)       # n_patches x d
        hidden = @encoder.run(tokens)              # contextual patches
        raw = T.matmul(hidden, @proj_out).map { |r| r[0] }
        # z-score across patches so the rendered image always keeps contrast
        mean = raw.sum / raw.length
        var = raw.sum { |v| (v - mean)**2 } / raw.length
        std = Math.sqrt(var) + 1e-9
        outs = raw.map { |v| sigmoid((v - mean) / std) }
        unpatch(outs)
      end

      # Render a grid (values in [0,1]) as ASCII shades.
      def self.ascii(grid_vals)
        shades = " .:-=+*#%@"
        grid_vals.map do |row|
          row.map { |v| shades[(v.clamp(0.0, 1.0) * (shades.length - 1)).round] }.join
        end.join("\n")
      end

      private

      def patchify(image)
        rows = []
        @per_side.times do |py|
          @per_side.times do |px|
            vec = []
            @patch.times do |dy|
              @patch.times do |dx|
                vec << image[py * @patch + dy][px * @patch + dx]
              end
            end
            rows << vec
          end
        end
        rows
      end

      def unpatch(outs)
        out = Array.new(@grid) { Array.new(@grid, 0.0) }
        idx = 0
        @per_side.times do |py|
          @per_side.times do |px|
            val = outs[idx]
            @patch.times do |dy|
              @patch.times do |dx|
                out[py * @patch + dy][px * @patch + dx] = val
              end
            end
            idx += 1
          end
        end
        out
      end

      def sigmoid(x)
        1.0 / (1.0 + Math.exp(-x))
      end
    end
  end
end

File: apps/core/src/main/java/bada/mind/Vision.java
package bada.mind;

/**
 * Image-processing Transformer (ViT-style) on the manifold-gauge Encoder. Cuts a
 * grid into patches, mixes them, and reconstructs the "mental image". Java port
 * of Bada::Transformer::Vision.
 */
public final class Vision {
    private final Encoder encoder;
    private final int grid, patch, perSide, patchDim;
    private final double[][] projIn;
    private final double[][] projOut;

    public Vision(Encoder encoder, int grid, int patch, long seed) {
        this.encoder = encoder;
        this.grid = grid;
        this.patch = patch;
        this.perSide = grid / patch;
        this.patchDim = patch * patch;
        Tensor.PRNG rng = new Tensor.PRNG(seed);
        this.projIn = Tensor.randmat(patchDim, encoder.dModel(), rng, 0.5);
        this.projOut = Tensor.randmat(encoder.dModel(), 1, rng, 0.5);
    }

    public double[][] process(double[][] image) {
        double[][] patches = patchify(image);
        double[][] tokens = Tensor.matmul(patches, projIn);
        double[][] hidden = encoder.run(tokens);
        double[][] proj = Tensor.matmul(hidden, projOut);
        double[] raw = new double[proj.length];
        for (int i = 0; i < proj.length; i++) raw[i] = proj[i][0];
        // z-score across patches so the rendered image always keeps contrast
        double mean = 0.0;
        for (double v : raw) mean += v;
        mean /= raw.length;
        double var = 0.0;
        for (double v : raw) var += (v - mean) * (v - mean);
        double std = Math.sqrt(var / raw.length) + 1e-9;
        double[] outs = new double[raw.length];
        for (int i = 0; i < raw.length; i++) outs[i] = sigmoid((raw[i] - mean) / std);
        return unpatch(outs);
    }

    public static String ascii(double[][] gridVals) {
        String shades = " .:-=+*#%@";
        StringBuilder sb = new StringBuilder();
        for (int y = 0; y < gridVals.length; y++) {
            for (double v : gridVals[y]) {
                int idx = (int) Math.round(Math.max(0.0, Math.min(1.0, v)) * (shades.length() - 1));
                sb.append(shades.charAt(idx));
            }
            if (y < gridVals.length - 1) sb.append('\n');
        }
        return sb.toString();
    }

    private double[][] patchify(double[][] image) {
        double[][] rows = new double[perSide * perSide][patchDim];
        int r = 0;
        for (int py = 0; py < perSide; py++) {
            for (int px = 0; px < perSide; px++) {
                int c = 0;
                for (int dy = 0; dy < patch; dy++)
                    for (int dx = 0; dx < patch; dx++)
                        rows[r][c++] = image[py * patch + dy][px * patch + dx];
                r++;
            }
        }
        return rows;
    }

    private double[][] unpatch(double[] outs) {
        double[][] out = new double[grid][grid];
        int idx = 0;
        for (int py = 0; py < perSide; py++) {
            for (int px = 0; px < perSide; px++) {
                double val = outs[idx++];
                for (int dy = 0; dy < patch; dy++)
                    for (int dx = 0; dx < patch; dx++)
                        out[py * patch + dy][px * patch + dx] = val;
            }
        }
        return out;
    }

    private static double sigmoid(double x) { return 1.0 / (1.0 + Math.exp(-x)); }
}
